platform/darwin/src/collator.mm

```diff
@@ -0,0 +1,89 @@
+#include <mbgl/style/expression/collator.hpp>
+#include <mbgl/text/language_tag.hpp>
+
+#include <sstream>
+
+#import <Foundation/Foundation.h>
+
+namespace mbgl {
+
+std::string localeIdentifier(const std::string& bcp47) {
+    // Based on Apple docs at:
+    // https://developer.apple.com/library/content/documentation/MacOSX/Conceptual/BPInternational/LanguageandLocaleIDs/LanguageandLocaleIDs.html#//apple_ref/doc/uid/10000171i-CH15-SW9
+    LanguageTag languageTag = LanguageTag::fromBCP47(bcp47);
+    std::stringstream localeIdentifier;
+    if (!languageTag.language) {
+        return localeIdentifier.str();
+    } else {
+        localeIdentifier << *(languageTag.language);
+    }
+
+    if (languageTag.script) {
+        localeIdentifier << "-" << *(languageTag.script);
+    }
+
+    if (languageTag.region) {
+        // TODO: Does locale identifier support UN M.49? If not we need to do a conversion or document
+        localeIdentifier << "_" << *(languageTag.region);
+    }
+
+    return localeIdentifier.str();
+}
+
+namespace style {
+namespace expression {
+
+class Collator::Impl {
+public:
+    Impl(bool caseSensitive, bool diacriticSensitive, optional<std::string> locale_)
+    : options((caseSensitive ? 0 : NSCaseInsensitiveSearch) |
+              (diacriticSensitive ? 0 : NSDiacriticInsensitiveSearch))
+    , locale(locale_ ?
+                [[NSLocale alloc] initWithLocaleIdentifier:@((*locale_).c_str())] :
+                [NSLocale currentLocale])
```

Comment:
`+[NSLocale currentLocale]` and `+[NSLocale autoupdatingCurrentLocale]` are based on the the system language, which may differ from the application’s interface language in the case where the application is unavailable in that language. The iOS and macOS map SDKs instead pass the return value of `+[NSLocale preferredLanguages]`, which reflects [the system’s preferred language setting](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/SpecifyingPreferences/SpecifyingPreferences.html#//apple_ref/doc/uid/10000171i-CH12), into `+[NSBundle preferredLocalizationsFromArray:forPreferences:]`, which matches it against the application’s available languages to produce a language code that the user prefers and that the application is localized into. The SDKs use this matching code for automatically localizing labels, as well as for some minor functionality like title casing and label transliteration in VoiceOver.

mapbox/mapbox-gl-js#6197 is meant to better reflect the potential discrepancy between the application’s current language and the overall system’s current language. In the meantime, `+currentLocale` is probably better than `+autoupdatingCurrentLocale`, given that #11660 hasn’t been implemented yet.

Reply:
So... just to summarize, for now `+currentLocale` is good, in the future we'll want to explore using `preferredLocalizationsFromArray` and tie into bundled localization information?

Reply:
Yep, thanks for reading my longwinded discourse. 😄